### PR TITLE
Added test case for importing a class from a private submodule into a public one

### PR DESCRIPTION
**Pull request by astrofrog**
Investigating the issue from https://github.com/astropy/astropy/pull/16974

### CODE REVIEW

**Comment by astrofrog** [comment]
For now this just adds the regression test. One way to solve that specific use case is to do:

```
.. automodapi:: sphinx_automodapi.tests.example_module.public
    :allowed-package-names: sphinx_automodapi.tests.example_module._private
```

however this could get tedious if we ever did use this pattern of private submodules extensively in a package. I wonder if actually what we want to do is to have an option to respect ``__all__`` if it is present, as the authoritative version of what should be documented in a module.

**Comment by astrofrog** [comment]
I think I found the solution - the issue is that currently we were not trusting __all__ and still excluding items from all that were imported from another location unless ``:allowed-package-names:`` was present. I think this doesn't make sense - I'm pretty sure that most people documenting a module that contains ``__all__`` will be expecting all items in ``__all__`` to be included in the API docs, so I've made it so that the ``onlylocals`` option is ignored if ``__all__`` is present.

File: sphinx_automodapi/tests/example_module/public.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+from ._private import Camelot
```
Comment (bsipocz) [inline sphinx_automodapi/tests/example_module/public.py:1]: Can we have a test where we import from a non-private module? I mean those cases work, so I just want to make sure that keeps being the case.
(although I see cases where they listed in multiple `__all__`s, probably those should be cleaned up, and (which is beyond scope here) preferably errored out by sphinx-automodapi?)

**Comment by nstarman** [comment]
I'm not familiar with the internals of this package. I know it uses ``autosummary`` and has templates. 
From that info, I would expect ``autosummary`` to generate the stub files via ``autodoc``

> A [autosummary](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html?highlight=autosummary#directive-autosummary) directive also generates short “stub” files for the entries listed in its content. These files by default contain only the corresponding [sphinx.ext.autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#module-sphinx.ext.autodoc) directive, but can be customized with templates. [(link)](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html)

The relevant ``autodoc`` directive would then have the [``:members:``](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-option-automodule-members) option which naturally uses ``__all__``.

Is that what's happening here?

File: sphinx_automodapi/tests/cases/public_from_private/output/api/sphinx_automodapi.tests.example_module.public.Camelot.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Camelot
+=======
+
+.. currentmodule:: sphinx_automodapi.tests.example_module.public
+
+.. autoclass:: Camelot
+   :show-inheritance:
```
Comment (astrofrog) [inline sphinx_automodapi/tests/cases/public_from_private/output/api/sphinx_automodapi.tests.example_module.public.Camelot.rst:7]: @nstarman we actually generate the stub files ourselves and they contain individual eg autoclass calls so the fix has to be and is in our code.

**Comment by astrofrog** [comment]
In future we might want to re examine using Sphinx directly to generate all the API docs (this package predates the ability of Sphinx to generate stub files)

File: sphinx_automodapi/tests/example_module/public.py (hunk 1 above)
Comment (astrofrog) [reply sphinx_automodapi/tests/example_module/public.py:1]: Sounds good, will adjust the test

**Comment by bsipocz** [comment]
> In future we might want to re examine using Sphinx directly to generate all the API docs (this package predates the ability of Sphinx to generate stub files)

I really wondered about this, too as a lot has happened in the past 10 years on sphinx land...

File: sphinx_automodapi/tests/cases/public_from_private/output/api/sphinx_automodapi.tests.example_module.public.Camelot.rst (hunk 2 above)
Comment (nstarman) [reply sphinx_automodapi/tests/cases/public_from_private/output/api/sphinx_automodapi.tests.example_module.public.Camelot.rst:7]: Then all LGTM. The results files seem reasonable :)

**Comment by astrofrog** [comment]
@bsipocz - I adjusted the test as you requested to include a class imported from a public submodule

**Comment by pllim** [comment]
Re: Using Sphinx directly -- We have issue at https://github.com/astropy/sphinx-automodapi/issues/147

**Review by bsipocz — APPROVED** [review]
Looks good now, thanks @astrofrog!

**Comment by pllim** [comment]
Thanks, all! @astrofrog , do you plan to do a release here or you need help?

**Comment by astrofrog** [comment]
Already done! (v0.18)

**Pull request merged**